### PR TITLE
Introduce python proto and update README 

### PR DESCRIPTION
**Pull request by janfeitsma**
_Q: Jurge, is this OK for added cmake complexity? I think we need to find help for improving cmake/codegen._ 

Objective: to automatically create python protobuf bindings for each component.

(Later use case: also create pybind11 bindings so components can be called from python.)

Still to do:
* codegen - DONE
* demo python script - DONE
* rename all dashes in component names to underscores because python protoc cannot handle them, ~causing build to fail at the moment~ -- DONE, sorry for the very large amount of changed files, burying the real changes.

Demo:

See the [demo script](https://github.com/janfeitsma/MRA-prototype/blob/pyproto/components/falcons/velocity_control/test/demo_protobuf.py).

To run new tests: `MRA-build.py -t -s velocity_control -d -j 1`

image

### CODE REVIEW

File: MRA-build.py [hunk 1]
```diff
@@ -104,7 +105,7 @@ def resolve_scope(iscope : str) -> list:
     for s in iscope.split(','):
         candidates = [c for c in components if c.endswith(s)]
         if len(candidates) == 1:
-            result.add(candidates[0] + ':all')
+            result.add(candidates[0] + '/...')
```
Comment (jveijck) [inline MRA-build.py]: what does '/.../' mean ? Perhaps comment to explain what branch means

File: MRA-codegen.py [hunk 2]
```diff
@@ -39,11 +39,11 @@
...
 
 def component_name_underscore(component: str) -> str:
-    """Mangle (relative) component name. Example: falcons/getball-intercept -> FALCONS_GETBALL_INTERCEPT"""
+    """Mangle (relative) component name. Example: falcons/getball_intercept -> FALCONS_GETBALL_INTERCEPT"""
     return component.replace('/', '_').replace('-', '_').upper()
 def component_name_camelcase(component: str) -> str:
-    """Mangle (relative) component name. Example: falcons/getball-intercept -> FalconsGetballIntercept"""
-    return ''.join([s.capitalize() for s in re.split('/|-', component)])
+    """Mangle (relative) component name. Example: falcons/getball_intercept -> FalconsGetballIntercept"""
+    return ''.join([s.capitalize() for s in re.split('/|-|_', component)])
```
Comment (jveijck) [inline MRA-codegen.py:46]: minus sign still supported ?

File: MRA-codegen.py [hunk 3]
```diff
@@ -70,16 +70,25 @@ def __init__(self, component: str, verbosity: int = DEFAULT_VERBOSE_LEVEL):
...
         self.analyze()
 
+    def check_component_name_valid(self):
+        """Ensure component name is valid, for instance no dash character ... (causes problem with protoc --python_out)"""
+        if '-' in self.component:
+            raise Exception(f'illegal component name, no dash (-) allowed in {self.component}')
+
+    def check_component_interface_valid(self):
+        """Ensure component interface (.proto files) are valid."""
+        # TODO: warn in case other than all_interface_parts .proto files are detected?
```
Comment (jveijck) [inline MRA-codegen.py:80]: Perhaps check if other files are in list ["Local", "State", "Parameters"]. If not matching, then unexpected proto file.

File: WORKSPACE.bazel [hunk 4]
```diff
@@ -9,10 +9,26 @@ http_archive(
...
     urls = ["https://github.com/protocolbuffers/protobuf/archive/refs/tags/v3.19.4.zip"],
     # note: on Ubuntu20, opencv DNN libraries depend on an older conflicting protobuf version
+    # note: py_proto_library might spam warnings:
+    #    "com_google_protobuf/python: warning: directory does not exist"
+    #    -> require a fix from 22Q1 - https://github.com/protocolbuffers/protobuf/issues/6049
 )
 load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
 protobuf_deps()
 
+# GRPC v1.42, for proto rules.
```
Comment (jveijck) [inline WORKSPACE.bazel:20]: perhaps add link to main page of the liberaries ? 
GRPC -> https://grpc.io/
Protobuf -> https://protobuf.dev/
nlohmann JSON -> https://github.com/nlohmann/json
gtest -> https://github.com/google/googletest

**Review by jveijck — APPROVED** [review]

File: MRA-build.py (hunk 1 above)
Comment (janfeitsma) [reply MRA-build.py]: solved by defining and making use of:
`BAZEL_ALL = '...' # see bazel syntax / cheatsheet`

File: MRA-codegen.py (hunk 2 above)
Comment (janfeitsma) [reply MRA-codegen.py:46]: No, see `check_component_name_valid` and the exception it would raise.

Comment (janfeitsma) [reply MRA-codegen.py:46]: Is this a problem for you? It just doesn't work with the python side of things ...

File: MRA-codegen.py (hunk 3 above)
Comment (janfeitsma) [reply MRA-codegen.py:80]: Well yeah that is what the TODO suggests... 

Gonna leave as is.

File: MRA-codegen.py (hunk 2 above)
Comment (janfeitsma) [reply MRA-codegen.py:46]: Gonna leave this function as is; it technically should focus on mangling, not reflect rules/choices elsewhere.

Comment (jveijck) [reply MRA-codegen.py:46]: agree

File: MRA-codegen.py (hunk 3 above)
Comment (jveijck) [reply MRA-codegen.py:80]: agree with proposal to leave as it is.

**Pull request merged**